### DISCUSSION TITLE
extend inventory model

### DISCUSSION THREAD

**Discussion by bex-ayoub** [medusajs/medusa#7983] [Q&A]
hi there!

im a bit new to Medusa, and I went through the docs and followed recipes for different ideas.
However, I need to extend the Inventory model itself and as I see there is no such model in Medusa or I didn't find it :")
any help?
in fact, I am trying to build a marketplace and we wanna keep the Product entity visible for any vendor, but the inventory items only visible for the store of the current user.

thnx

**Comment by bex-ayoub** [comment]
any help here? :")
thnx

***
to better describe the situation, we want to build a marketplace for our different vendors.
I followed [this guide](https://medusajs.com/blog/extending-medusa-usecase-marketplace/) while aligning with the new recipe [here](https://docs.medusajs.com/recipes/marketplace)
(I also followed the guided tuto [here](https://blog.perseides.org/marketplace-lets-follow-the-recipe))
but we actually want to disconnect the Product from the store, because the same Product can exist for multiple stores.
So, we need to link the Product Variant & Inventory Item with the Store (i.e vendor).
however, I didn't find a way to extend the Inventory model.

**Reply by okechukwu0127** [reply]
@bex-ayoub I am also facing this same issue of extending the Inventory Service as it was installed from @medusajs/inventory

But i have been able to extend the product service to allow vendors see only products from there own store.
also extended the user model to include store_id

so when a user creates a new product, the product is saved with the store_id value from the user object

I was able to achieve this by adding store_id to the product model then extending the productService to only show products who's store_id is same with the store_id from the user object

**Reply by bex-ayoub** [reply]
I was able to extend Inventory
as so:

```
import { InventoryItem as MedusaInventoryItem } from "@medusajs/inventory/dist/models";
import { Column, Entity } from "typeorm";

@Entity()
export class InventoryItem extends MedusaInventoryItem {
  // linking InventoryItem to the Pos-Connection which presents the POS system
  @Column()
  pos_connection_id: string;
}

```

**Reply by imash96** [reply]
@bex-ayoub are you talking about medusa v1.x aur medusa v2.0

**Reply by bex-ayoub** [reply]
> @bex-ayoub are you talking about medusa v1.x aur medusa v2.0

hey @imash96 
it's v1

**Reply by imash96** [reply]
Any idea how can we exdtend core entity property in medusa v2. Like I want to excent product module to have properties like brand, gender etc..

**Reply by okechukwu0127** [reply]
> I was able to extend Inventory as so:
> 
> ```
> import { InventoryItem as MedusaInventoryItem } from "@medusajs/inventory/dist/models";
> import { Column, Entity } from "typeorm";
> 
> @Entity()
> export class InventoryItem extends MedusaInventoryItem {
>   // linking InventoryItem to the Pos-Connection which presents the POS system
>   @Column()
>   pos_connection_id: string;
> }
> ```


Thanks for the update, it was really insightfull to create a new model entity.
I was also able to create a migration for the new field

I was able to create an entity, create  and run migration successfully, **but nothing changed when i went to the inventory screen. I still see the same reocrds**.


src/models/inventory-item.ts
`

      import { InventoryItem as MedusaInventoryItem } from "@medusajs/inventory/dist/models";
    import { Store } from "@medusajs/medusa";
    import { Column, Entity, Index, JoinColumn, ManyToOne } from "typeorm";

    @Entity()
    export class InventoryItem extends MedusaInventoryItem {
      @Index("inventory_itemStoreId")
      @Column({ nullable: true })
      store_id?: string;

      @ManyToOne(() => Store, (store) => store.members)
      @JoinColumn({ name: "store_id", referencedColumnName: "id" })
      store?: Store;
    }

`

**WHAT I WAS EXPECTING**
I wanted the logged in user to see only inventories related to his store.
When a new inventory is created, it should be created with the logged in user store_id


<img width="1172" alt="image" src="https://github.com/user-attachments/assets/4cf4de3d-47dd-4c16-b20d-bb2fd42ab68d">


Let me show you what my code looks like

You can add me  on skype for a quick chat
https://join.skype.com/invite/jyJxARQShO5o


relative path to the core InventoryItemService file i extended node_modules/@medusajs/inventory/dist/services/inventory-item.d.ts

src/services/inventory-item.ts
`

          import { InventoryItemService as MedusaInventoryItemService } from "@medusajs/inventory/dist/services";
          import { User } from "@medusajs/medusa";
          import { Lifetime } from "awilix";
          import ProductRepository from "../repositories/product";
          import { CreateInventoryItemInput } from "@medusajs/types/dist/inventory/mutations";
          import { FindConfig, SharedContext } from "@medusajs/types";
          import {
          FilterableInventoryItemProps,
          InventoryItemDTO,
        } from "@medusajs/types/dist/inventory/bundle";
        import { InventoryItem } from "src/models/inventory-item";

        type CreateInventoryItem = {
          store_id?: string;
        } & CreateInventoryItemInput[];

        type InventoryItemSelector = {
          store_id?: string;
        } & FilterableInventoryItemProps;

        class InventoryItemService extends MedusaInventoryItemService {
          static LIFE_TIME = Lifetime.SCOPED;

          protected readonly loggedInUser_: User | null;
          protected readonly productRepository_: typeof ProductRepository;

          constructor(container, options) {
            // @ts-expect-error prefer-rest-params
            super(...arguments);

            try {
              this.loggedInUser_ = container.loggedInUser;
            } catch (e) {
              // avoid errors when backend first runs
            }
          }

          async list(
            selector: InventoryItemSelector,
            config?: FindConfig<InventoryItem>,
            context?: SharedContext
          ): Promise<InventoryItemDTO[]> {
            if (
              this.loggedInUser_?.role !== "admin" &&
              !selector.store_id &&
              this.loggedInUser_?.store_id
            ) {
              selector.store_id = this.loggedInUser_.store_id;
            }

            config.select?.push("store_id");

            config.relations?.push("store");

            return await super.list(selector, config);
          }

          async create(
            data: CreateInventoryItem,
            context?: SharedContext
          ): Promise<InventoryItem[]> {
            if (!data.store_id && this.loggedInUser_?.store_id) {
              data.store_id = this.loggedInUser_.store_id;
            }

            return await super.create(data, context);
          }
        }

        export default InventoryItemService;

  
`

**Reply by SalahAdDin** [reply]
Why was this closed?
